pkg/kubelet/dockershim/checkpoint_store.go

```diff
@@ -0,0 +1,85 @@
+package dockershim
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// CheckpointStore provides the interface for checkpoint storage backend
+type CheckpointStore interface {
+	Add(key string, data []byte) error
+	Get(key string) ([]byte, error)
+	Delete(key string) error
+	List() ([]string, error)
+}
+
+// FileStore is an implementation of CheckpointStore interface which stores checkpoint in file.
+type FileStore struct {
+	path string
+}
+
+func (fstore *FileStore) Add(key string, data []byte) error {
+	if _, err := os.Stat(fstore.path); os.IsNotExist(err) {
+		if err = os.MkdirAll(fstore.path, 755); err != nil {
```

Comment:
if this is used concurrently, you might see IsNotExist() and then fail MkdirAll() with "already exists".  You should either hold a mutex or else handle errors better - maybe check os.Stat() again in case of error, and if the dir exists now, assume you lost the race.